tests/Credentials/CredentialProviderTest.php

```diff
@@ -664,6 +664,167 @@ public function testEnsuresSourceProfileIsSpecified()
         }
     }
 
+    public function testAssumeRoleInConfigFromCredentialsSourceFails()
```

Comment:
* Can you add a case where the user sets both credential_source and source_profile?
* Is there a way to mock the success case for Ec2InstanceMetadata and EcsContainer? Ideally there'd be a fail and success case for each type.

Reply:
No unfortunately not, I tried for a while to get it to work.  The reason it won't work is because of the fact that you can't inject a mock into this function.  Take the ecs for instance.  Given that it is a static method that only takes a config, you can't inject a mock in any way:
    public static function ecsCredentials(array $config = [])
    {
        return new EcsCredentialProvider($config);
    }

Reply:
Makes sense. More invasive measures could be taken by adding mock functions to replace `instanceProfile()` and the like, but at that point you'd only be testing the method is called, which is already being tested via the expected exception message here.